### PR TITLE
Add opensearch detail dashboard

### PR DESCRIPTION
**Pull request by Errahulaws**
### Checklist
- [ ] Issue(s) linked

### CODE REVIEW

File: dashboards/opensearch/opensearch_details.sp [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+locals {
+  opensearch_common_tags = {
+    service = "AWS/OpenSearch"
+  }
+}
+
+# dashboard.tf
+
+dashboard "opensearch_domain_detail" {
+  title         = "OpenSearch Domain Detail"
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:12]:
```suggestion
  title         = "AWS OpenSearch Domain Detail"
```

File: dashboards/opensearch/opensearch_details.sp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+  tags = merge(local.opensearch_common_tags, {
+    type = "Detail"
+  })
+
+  input "opensearch_domain_name" {
+    title = "Select an OpenSearch domain:"
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:19]:
```suggestion
    title = "Select a domain:"
```

File: dashboards/opensearch/opensearch_details.sp [hunk 3]
```diff
@@ -0,0 +1,223 @@
+# locals.tf
+
+locals {
+  opensearch_common_tags = {
+    service = "AWS/OpenSearch"
+  }
+}
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:7]: Can you please check the detail pages of the other services and incorporate them into the dashboards/opensearch/opensearch.sp file to ensure consistency? Thanks.

File: dashboards/opensearch/opensearch_details.sp [hunk 4]
```diff
@@ -0,0 +1,223 @@
+# locals.tf
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:1]:
```suggestion
```

File: dashboards/opensearch/opensearch_details.sp [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+locals {
+  opensearch_common_tags = {
+    service = "AWS/OpenSearch"
+  }
+}
+
+# dashboard.tf
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:9]:
```suggestion
```

File: dashboards/opensearch/opensearch_details.sp (hunk 3 above)
Comment (Errahulaws) [reply dashboards/opensearch/opensearch_details.sp]: @khushboo9024  thanks 
Its updated

**Review by khushboo9024 — COMMENTED** [review]
Hello [Errahulaws](https://github.com/Errahulaws), thank you for submitting this pull request. We have reviewed the code and made adjustments to accurately connect the relationship graph. I've shared the preliminary changes we implemented locally for your reference. Please use this as a guide and continue with the necessary modifications.

Note: 
- We are using the domain ARN as the input value, not the domain name; this was the reason why the edges were not linking
- The existing WHERE clause in the opensearch_domain node is incorrect, which was the reason the OpenSearch domain icon did not appear in the relationship graph.
```
opensearch_details.sp

dashboard "opensearch_domain_detail" {
  title = "OpenSearch Domain Detail"

  tags = merge(local.opensearch_common_tags, {
    type = "Detail"
  })

  input "opensearch_arn" {
    title = "Select a domain:"
    query = query.opensearch_domain_input
    width = 4
  }

  container {

    card {
      width = 2
      query = query.opensearch_domain_instance_type
      args  = [self.input.opensearch_arn.value]
    }

    card {
      width = 2
      query = query.opensearch_domain_version
      args  = [self.input.opensearch_arn.value]
    }

    card {
      width = 2
      query = query.opensearch_domain_endpoint
      args  = [self.input.opensearch_arn.value]
    }
  }

  with "vpc_security_groups_for_opensearch" {
    query = query.vpc_security_groups_for_opensearch
    args  = [self.input.opensearch_arn.value]
  }

  with "vpc_subnet_for_opensearch" {
    query = query.vpc_subnet_ids_for_opensearch
    args  = [self.input.opensearch_arn.value]
  }

  container {

    graph {

      title     = "Relationships"
      type      = "graph"
      direction = "TD"

      node {
        base = node.opensearch_domain
        args = {
          opensearch_arns = [self.input.opensearch_arn.value]
        }
      }

      node {
        base = node.vpc_security_group
        args = {
          vpc_security_group_ids = with.vpc_security_groups_for_opensearch.rows[*].security_group_id
        }
      }

      node {
        base = node.vpc_subnet
        args = {
          vpc_subnet_ids = with.vpc_subnet_for_opensearch.rows[*].subnet_id
        }
      }

      edge {
        base = edge.opensearch_domain_to_vpc_security_group
        args = {
          opensearch_arn = self.input.opensearch_arn.value
        }
      }

      edge {
        base = edge.opensearch_domain_to_vpc_subnet
        args = {
          opensearch_arn = self.input.opensearch_arn.value
        }
      }
    }
  }

  container {
    width = 12

    table {
      title = "Overview"
      type  = "line"
      width = 6
      query = query.opensearch_domain_overview
      args  = [self.input.opensearch_arn.value]
    }

    table {
      title = "Tags"
      width = 6
      query = query.opensearch_domain_tags
      args  = [self.input.opensearch_arn.value]
    }
  }

}

query "opensearch_domain_input" {
  sql = <<-EOQ
    SELECT
      domain_name AS label,
      arn AS value,
      json_build_object(
        'domain_name', domain_name,
        'arn', arn
      ) AS tags
    FROM
      aws_opensearch_domain
    ORDER BY
      domain_name;
  EOQ
}

query "opensearch_domain_tags" {
  sql = <<-EOQ
    SELECT
      tags->>'Key' AS "Key",
      tags->>'Value' AS "Value"
    FROM (
      SELECT jsonb_array_elements(tags_src) AS tags
      FROM aws_opensearch_domain
      WHERE arn = $1
    ) AS subquery;
  EOQ
}


query "opensearch_domain_version" {
  sql = <<-EOQ
    SELECT
      engine_version as "Engine Version"
    FROM
      aws_opensearch_domain
    WHERE
      arn = $1;
  EOQ
}

query "opensearch_domain_instance_type" {
  sql = <<-EOQ
    SELECT
      cluster_config->>'InstanceType' AS "Instance Type"
    FROM
      aws_opensearch_domain
    WHERE
      arn = $1;
  EOQ
}


query "opensearch_domain_overview" {
  sql = <<-EOQ
    SELECT
      domain_name AS "Domain Name",
      arn AS "ARN",
      created as "Created",
      processing as "Processing",
      region as "Region",
      account_id as "Account ID"
    FROM
      aws_opensearch_domain
    WHERE
      arn = $1;
  EOQ
}

query "opensearch_domain_endpoint" {
  sql = <<-EOQ
    SELECT
      endpoint as "Endpoint"
    FROM
      aws_opensearch_domain
    WHERE
      arn = $1;
  EOQ
}

query "vpc_security_groups_for_opensearch" {
  sql = <<-EOQ
    SELECT
      jsonb_array_elements_text(vpc_options -> 'SecurityGroupIds') AS security_group_id
    FROM
      aws_opensearch_domain
    WHERE
      arn = $1;
  EOQ
}

query "vpc_subnet_ids_for_opensearch" {
  sql = <<-EOQ
    SELECT
      jsonb_array_elements_text(vpc_options -> 'SubnetIds') AS subnet_id
    FROM
      aws_opensearch_domain
    WHERE
      arn = $1;
  EOQ
}
```
```
opensearch_edges.sp

edge "opensearch_domain_to_vpc_security_group" {
  title = "Security Groups"
  sql   = <<-EOQ
    SELECT
      arn as from_id,
      jsonb_array_elements_text(vpc_options -> 'SecurityGroupIds') AS to_id
    FROM
      aws_opensearch_domain
      JOIN jsonb_array_elements(vpc_options -> 'SecurityGroupIds') AS s ON true
    WHERE
      arn = $1;
  EOQ
  param "opensearch_arn" {}
}


edge "opensearch_domain_to_vpc_subnet" {
  title = "subnet"

  sql = <<-EOQ
    SELECT
      jsonb_array_elements_text(vpc_options -> 'SecurityGroupIds') as from_id,
      jsonb_array_elements_text(vpc_options -> 'SubnetIds') AS to_id
    FROM
      aws_opensearch_domain
      JOIN jsonb_array_elements(vpc_options -> 'SecurityGroupIds') AS s ON true
    WHERE
      arn = $1;
  EOQ

  param "opensearch_arn" {}
}
```
```
opensearch_nodes.sp

node "opensearch_domain" {
  category = category.opensearch_domain

  sql = <<-EOQ
    select
      arn as id,
      title as title,
      jsonb_build_object(
        'ARN', arn,
        'Domain ID', domain_id,
        'Domain Name', domain_name,
        'Engine Version', engine_version,
        'Account ID', account_id,
        'Region', region
      ) as properties
    from
      aws_opensearch_domain
    where
      arn = any($1);
  EOQ

  param "opensearch_arns" {}
}
```

Please let us know if any other details are required.
Thanks.

**Comment by Errahulaws** [comment]
@khushboo9024 Thanks I have made changes I will update the PR 
Please review and merge as needed

**Comment by Errahulaws** [comment]
@khushboo9024 Done with the changes 
can we have this review and merged

File: dashboards/opensearch/opensearch_details.sp [hunk 6]
```diff
@@ -0,0 +1,211 @@
+dashboard "opensearch_domain_detail" {
+  title = "OpenSearch Domain Detail"
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:2]:
```suggestion
  title = "AWS OpenSearch Domain Detail"
```

File: dashboards/opensearch/opensearch_details.sp [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+      domain_name AS label,
+      arn AS value,
+      json_build_object(
+        'domain_name', domain_name,
+        'arn', arn
+      ) AS tags
+    FROM
+      aws_opensearch_domain
+    ORDER BY
+      domain_name;
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:123]:
```suggestion
    select
      domain_name as label,
      arn as value,
      json_build_object(
        'domain_name', domain_name,
        'arn', arn
      ) as tags
    from
      aws_opensearch_domain
    order by
      domain_name;
```
@Errahulaws  can you please format all the queries as above.

File: dashboards/opensearch/opensearch_nodes.sp [hunk 8]
```diff
@@ -21,3 +21,28 @@ node "opensearch_domain" {
...
+        'Region', region
+      ) as properties
+    from
+      aws_opensearch_domain
+    where
+      arn = any($1);
+  EOQ
+
+  param "opensearch_arns" {}
+}
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_nodes.sp:48]: This node(opensearch_domain_arn) is not required we can use the existing node(opensearch_domain), we just need to correct the where clause. Can you please update node `opensearch_domain` as below and remove node `opensearch_domain_arn`
```
node "opensearch_domain" {
  category = category.opensearch_domain
  sql = <<-EOQ
    select
      arn as id,
      title as title,
      jsonb_build_object(
        'ARN', arn,
        'Domain ID', domain_id,
        'Domain Name', domain_name,
        'Engine Version', engine_version,
        'Account ID', account_id,
        'Region', region
      ) as properties
    from
      aws_opensearch_domain
    where
      arn = any($1);
  EOQ

  param "opensearch_arns" {}
}

```

File: dashboards/opensearch/docs/opensearch_details.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+This dashboard answers the following questions for each instance:
+
+- OpenSearch details
```
Comment (khushboo9024) [inline dashboards/opensearch/docs/opensearch_details.md:3]:
```suggestion
This dashboard answers the following questions for each domain:

- How is the domain configured?
- What tags are applied?
- What relationships does the function have with other resources?
```

**Review by khushboo9024 — COMMENTED** [review]
@Errahulaws  Can we please add more details, such as security groups and subnets associated with the domain? You can refer to the Lambda function detail page for an example. Please see the screenshot below.:
image

File: dashboards/opensearch/opensearch_details.sp [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+  input "opensearch_arn" {
+    title = "Select a domain:"
+    query = query.opensearch_domain_input
+    width = 4
+  }
+
+  container {
+
+    card {
+      width = 2
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:17]:
```suggestion
      width = 3
```

File: dashboards/opensearch/opensearch_details.sp [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+      query = query.opensearch_domain_instance_type
+      args  = [self.input.opensearch_arn.value]
+    }
+
+    card {
+      width = 2
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:23]:
```suggestion
      width = 3
```

File: dashboards/opensearch/opensearch_details.sp [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+      query = query.opensearch_domain_version
+      args  = [self.input.opensearch_arn.value]
+    }
+
+    card {
+      width = 2
```
Comment (khushboo9024) [inline dashboards/opensearch/opensearch_details.sp:29]:
```suggestion
      width = 3
```

**Comment by Errahulaws** [comment]
@khushboo9024 Updated requested changes 
Please review.

**Review by khushboo9024 — APPROVED** [review]

**Pull request merged**